traffic/traffic_info.cpp

```diff
@@ -85,8 +125,275 @@ TrafficInfo TrafficInfo::BuildForTesting(Coloring && coloring)
+        ow = false;
+        break;
+      }
+    }
+
+    auto const numDirs = ow ? 1 : 2;
+    size_t numSegsForThisFid = j - i;
+    CHECK_GREATER(numDirs, 0, ());
+    CHECK_EQUAL(numSegsForThisFid % numDirs, 0, ());
+    numSegsForThisFid /= numDirs;
+
+    fids.push_back(keys[i].m_fid);
+    numSegs.push_back(numSegsForThisFid);
+    oneWay.push_back(ow);
+
+    i = j;
+  }
+
+  MemWriter<vector<uint8_t>> memWriter(result);
+  WriteToSink(memWriter, kLatestKeysVersion);
+  WriteVarUint(memWriter, fids.size());
+
+  {
+    BitWriter<decltype(memWriter)> bitWriter(memWriter);
+
+    uint32_t prevFid = 0;
+    for (auto const & fid : fids)
+    {
+      uint32_t const fidDiff = fid - prevFid;
+      bool ok = coding::GammaCoder::Encode(bitWriter, fidDiff + 1);
+      ASSERT(ok, ());
+      prevFid = fid;
+    }
+
+    for (auto const & s : numSegs)
+    {
+      bool ok = coding::GammaCoder::Encode(bitWriter, s + 1);
+      ASSERT(ok, ());
+    }
+
+    for (auto const & val : oneWay)
+      bitWriter.Write(val ? 1 : 0, 1 /* numBits */);
+  }
+}
+
+// static
+void TrafficInfo::DeserializeTrafficKeys(vector<uint8_t> const & data,
+                                         vector<TrafficInfo::RoadSegmentId> & result)
+{
+  MemReader memReader(data.data(), data.size());
+  ReaderSource<decltype(memReader)> src(memReader);
+  auto const version = ReadPrimitiveFromSource<uint8_t>(src);
+  CHECK_EQUAL(version, kLatestKeysVersion, ("Unsupported version of traffic values."));
+  auto const n = static_cast<size_t>(ReadVarUint<uint64_t>(src));
+
+  vector<uint32_t> fids(n);
+  vector<size_t> numSegs(n);
+  vector<bool> oneWay(n);
+
+  {
+    BitReader<decltype(src)> bitReader(src);
+    uint32_t prevFid = 0;
+    for (size_t i = 0; i < n; ++i)
+    {
+      prevFid += coding::GammaCoder::Decode(bitReader) - 1;
+      fids[i] = prevFid;
+    }
+
+    for (size_t i = 0; i < n; ++i)
+      numSegs[i] = coding::GammaCoder::Decode(bitReader) - 1;
+
+    for (size_t i = 0; i < n; ++i)
+      oneWay[i] = bitReader.Read(1) > 0;
+  }
+
+  ASSERT_EQUAL(src.Size(), 0, ());
+
+  result.clear();
+  for (size_t i = 0; i < n; ++i)
+  {
+    auto const fid = fids[i];
+    uint8_t numDirs = oneWay[i] ? 1 : 2;
+    for (size_t j = 0; j < numSegs[i]; ++j)
+    {
+      for (uint8_t dir = 0; dir < numDirs; ++dir)
+      {
+        RoadSegmentId key(fid, j, dir);
+        result.push_back(key);
+      }
+    }
+  }
+}
+
+// static
+void TrafficInfo::SerializeTrafficValues(vector<SpeedGroup> const & values,
+                                         vector<uint8_t> & result)
+{
+  vector<uint8_t> buf;
+  MemWriter<vector<uint8_t>> memWriter(buf);
+  WriteToSink(memWriter, kLatestValuesVersion);
+  WriteVarUint(memWriter, values.size());
+  {
+    BitWriter<decltype(memWriter)> bitWriter(memWriter);
+    for (auto const & v : values)
+    {
+      uint8_t const u = static_cast<uint8_t>(v);
+      CHECK_LESS(u, (1 << 3), ("Speed group's value does not fit into 3 bits."));
```

Comment:
I still insist that you need to static_assert here that speed group count is at most 8, and check (or, better, assert, as it's cheaper) that v is not Count.